### PR TITLE
Jfoll/automate config api solution prod 10165

### PR DESCRIPTION
**Pull request by johnfolly**
# Test Commands

```bash
echo "Babylon"
babylon api solution get-all
babylon api solution get-all -f 'id,name'
babylon api solution get-all -o ../test.json
babylon api solution get-all -f 'id,name' -o ../test.json
export SOL=$(babylon api solution create -n "Babylon Solution" | grep -oh "\w*sol-\w*")
echo $SOL
babylon api solution get $SOL
babylon api solution update tests/api_files/Solution.yaml
babylon api solution get $SOL -f 'id,name'
babylon api solution get $SOL -o ../test.json
babylon api solution get $SOL -f 'id,name' -o ../test.json
babylon api solution get-current
babylon api solution get-current -f 'id,name'
babylon api solution get-current -o ../test.json
babylon api solution get-current -f 'id,name' -o ../test.json
babylon api solution delete $SOL -f
babylon api solution get-all -f 'id,name'
echo "Babylon"
babylon api solution get $SOL
babylon api solution update tests/api_files/Solution.yaml
babylon api solution get $SOL -f 'id,name'
babylon api solution get $SOL -o ../test.json
babylon api solution get $SOL -f 'id,name' -o ../test.json
babylon api solution get-current
babylon api solution get-current -f 'id,name'
babylon api solution get-current -o ../test.json
babylon api solution get-current -f 'id,name' -o ../test.json
babylon api solution delete $SOL -f
echo "Babylon"
babylon api solution create
```

### CODE REVIEW

File: Babylon/groups/api/groups/solution/commands/get_current.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        retrieved_solution = filter_api_response_item(retrieved_solution, fields.split(","))
+    if output_file:
+        converted_content = convert_keys_case(retrieved_solution, underscore_to_camel)
+        try:
+            with open(output_file, "w") as _f:
+                json.dump(converted_content, _f, ensure_ascii=False)
+        except TypeError:
+            with open(output_file, "w") as _f:
+                json.dump(converted_content.to_dict(), _f, ensure_ascii=False)
+        logger.debug(pformat(retrieved_solution))
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/get_current.py:79]:
```python
with open(output_file, "w") as _f:
  try:
    json.dump(...)
  except TypeError:
    json.dump(...)  
```

File: Babylon/groups/api/groups/solution/commands/get.py [hunk 2]
```diff
@@ -1,42 +1,85 @@
...
-@require_deployment_key("solution_id", "solution_id")
-@option("-o", "--output_file", "output_file", help="File to which content should be outputed (json-formatted)")
 @timing_decorator
-def get(solution_api: SolutionApi, solution_id: str, organization_id: str, output_file: Optional[str] = None):
-    """Get the state of the solution in the API"""
+@argument("solution_id", type=str)
+@require_deployment_key("organization_id", "organization_id")
+@option(
+    "-o",
+    "--output_file",
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/get.py:34]: avoid underscore in bash arguments `output-file`

File: Babylon/groups/api/groups/solution/commands/get.py [hunk 3]
```diff
@@ -1,42 +1,85 @@
...
         logger.error(f"Solution {solution_id} does not exists in organization {organization_id}.")
-    except cosmotech_api.exceptions.UnauthorizedException as _e:
+        return
+    except UnauthorizedException:
         logger.error("Unauthorized access to the cosmotech api")
+        return
+
+    if fields:
+        retrieved_solution = filter_api_response_item(retrieved_solution, fields.split(","))
+    if output_file:
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/get.py:72]:
```python
if not output_file:
  logger.info(...)
  return
```

File: Babylon/groups/api/groups/solution/commands/get.py [hunk 4]
```diff
@@ -1,42 +1,85 @@
...
+    "fields",
+    type=str,
+    help="Fields witch will be keep in response data, by default all",
+)
+def get(
+    solution_api: SolutionApi,
+    solution_id: str,
+    organization_id: str,
+    output_file: Optional[str] = None,
+    fields: str = None,
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/get.py:51]: type should be `Optional[str]`

File: Babylon/groups/api/groups/solution/commands/get_all.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    "--fields",
+    "fields",
+    type=str,
+    help="Fields witch will be keep in response data, by default all",
+)
+def get_all(
+    solution_api: SolutionApi,
+    organization_id: str,
+    output_file: Optional[str] = None,
+    fields: str = None,
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/get_all.py:48]: type should be `Optional[str]`

**Review by o-reo — CHANGES_REQUESTED** [review]
Thanks for this PR, I have some remarks.
I'll test usage and let you know

**Review by o-reo — APPROVED** [review]
LGTM

**Comment by o-reo** [comment]
Just tested the commands.
I think create should be able to output in a file so you can chain commands without using bash vars

File: Babylon/groups/api/groups/solution/commands/create.py [hunk 6]
```diff
@@ -1,49 +1,117 @@
...
+    help="New solution description",
+)
+@option(
+    "-s",
+    "--select",
+    "select",
+    type=bool,
+    help="Select this new Solution as babylon context solution ?",
+    default=True,
+)
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/create.py:63]: Add output file option

File: Babylon/groups/api/groups/solution/commands/create.py [hunk 7]
```diff
@@ -1,49 +1,117 @@
...
+    except UnauthorizedException:
+        logger.error("Unauthorized access to the cosmotech api")
+        return
+    except NotFoundException:
+        logger.error(f"Organization with id {organization_id} does not exists.")
+        return
+
+    if select:
+        env.configuration.set_deploy_var("solution_id", retrieved_solution["id"])
+
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/create.py:115]: Output the retrieved solution in a file if the output file option has been specified

File: Babylon/groups/api/groups/solution/commands/get.py [hunk 8]
```diff
@@ -1,42 +1,84 @@
...
 @command()
+@allow_dry_run
 @pass_solution_api
-@require_deployment_key("organization_id", "organization_id")
-@require_deployment_key("solution_id", "solution_id")
-@option("-o", "--output_file", "output_file", help="File to which content should be outputed (json-formatted)")
 @timing_decorator
-def get(solution_api: SolutionApi, solution_id: str, organization_id: str, output_file: Optional[str] = None):
-    """Get the state of the solution in the API"""
+@argument("solution_id")
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/get.py:31]: Add input file option to get ID from a file

File: Babylon/groups/api/groups/solution/commands/update.py [hunk 9]
```diff
@@ -1,56 +1,73 @@
...
 
 
 @command()
+@allow_dry_run
+@timing_decorator
 @pass_solution_api
-@argument("solution_file")
-@option("-e", "--use-solution-file", "use_solution_file", is_flag=True,
-        help="Should the path be in the solution ?")
+@argument("solution_file", type=str)
```
Comment (o-reo) [inline Babylon/groups/api/groups/solution/commands/update.py:27]: Should be of type Path

**Review by o-reo — CHANGES_REQUESTED** [review]
Sorry I see those while trying the commands

File: Babylon/groups/api/groups/solution/commands/get.py (hunk 8 above)
Comment (johnfolly) [reply Babylon/groups/api/groups/solution/commands/get.py:31]: I don't think that it is necessary.
If someone creates a solution for example if he wants to see the created solution details he will just open the output file generated by create commands.

But creating a new file, copying the id, setting the file format ... and then running the command is painful I think.
it's easier to get the list of all solutions, copy the id of the one which I'm interested in and run the command with.

Also `get` is not a part an deployment it's just an accessory commands

File: Babylon/groups/api/groups/solution/commands/update.py (hunk 9 above)
Comment (johnfolly) [reply Babylon/groups/api/groups/solution/commands/update.py:27]: solution file can't be an click.Path

```python
converted_solution_content = get_api_file(
        api_file_path=solution_file,
        use_working_dir_file=use_working_dir_file,
        logger=logger,
 )
```

use_working_dir_file allow to use file from babylon working directory witch is not necessary the current path location

File: Babylon/groups/api/groups/solution/commands/get.py (hunk 8 above)
Comment (o-reo) [reply Babylon/groups/api/groups/solution/commands/get.py:31]: There could be cases where the solution already exists and we want to read it to act on other parts of the infrastructure. Babylon is not just for deployment

File: Babylon/groups/api/groups/solution/commands/update.py (hunk 9 above)
Comment (o-reo) [reply Babylon/groups/api/groups/solution/commands/update.py]: my bad ;)

**Comment by o-reo** [comment]
Could you rebase to main and resolve conflict so we can merge ?

**Comment by johnfolly** [comment]
I rebased, there should be no more conflicts

**Pull request merged**